File: backend/test/unit/retryPolicy.test.ts
import { describe, expect, it } from 'vitest';
import {
  canRetrySameInstrument,
  hasSufficientSpacing,
  RETRY_POLICY,
} from '../../src/routing/retryPolicy.js';
import type { NormalizedDecline } from '../../src/domain/declines.js';

function decline(
  retryClass: NormalizedDecline['retryClass'],
): Pick<NormalizedDecline, 'retryClass'> {
  return { retryClass };
}

describe('canRetrySameInstrument — Non-negotiable #9', () => {
  it('refuses a hard (retryClass "never") decline outright, regardless of attempt count', () => {
    const result = canRetrySameInstrument(decline('never'), 0);
    expect(result.allowed).toBe(false);
    expect(result.reason).toMatch(/hard decline/i);
  });

  it('refuses a "review" (unmapped) decline outright', () => {
    const result = canRetrySameInstrument(decline('review'), 0);
    expect(result.allowed).toBe(false);
    expect(result.reason).toMatch(/review/i);
  });

  it('allows a retryable decline under the attempt cap', () => {
    const result = canRetrySameInstrument(decline('same_instrument_later'), 1);
    expect(result.allowed).toBe(true);
  });

  it('refuses a retryable decline once the attempt cap is reached', () => {
    const result = canRetrySameInstrument(
      decline('same_instrument_later'),
      RETRY_POLICY.maxAttemptsPerPayment,
    );
    expect(result.allowed).toBe(false);
    expect(result.reason).toMatch(/max attempts/i);
  });

  it('a hard decline is refused even with zero prior attempts and plenty of budget left', () => {
    // Guards against a bug where the attempt-cap check runs first and
    // masks the hard-decline check.
    const result = canRetrySameInstrument(decline('never'), 0, {
      maxAttemptsPerPayment: 100,
      minSpacingMs: 0,
    });
    expect(result.allowed).toBe(false);
  });

  it('retry_after_challenge and retry_different_psp are retryable (subject to the attempt cap)', () => {
    expect(canRetrySameInstrument(decline('retry_after_challenge'), 0).allowed).toBe(true);
    expect(canRetrySameInstrument(decline('retry_different_psp'), 0).allowed).toBe(true);
  });
});

describe('hasSufficientSpacing', () => {
  it('is false immediately after the last attempt', () => {
    const now = new Date('2026-01-01T00:00:00Z');
    expect(hasSufficientSpacing(now, now)).toBe(false);
  });

  it('is true once minSpacingMs has elapsed', () => {
    const last = new Date('2026-01-01T00:00:00.000Z');
    const now = new Date(last.getTime() + RETRY_POLICY.minSpacingMs);
    expect(hasSufficientSpacing(last, now)).toBe(true);
  });

  it('is false one millisecond before minSpacingMs has elapsed', () => {
    const last = new Date('2026-01-01T00:00:00.000Z');
    const now = new Date(last.getTime() + RETRY_POLICY.minSpacingMs - 1);
    expect(hasSufficientSpacing(last, now)).toBe(false);
  });
});

File: backend-go/internal/routing/retrypolicy_test.go
package routing

import (
	"regexp"
	"testing"
	"time"

	"github.com/alphapayments/payment-orchestrator/internal/domain"
)

func TestCanRetrySameInstrument(t *testing.T) {
	t.Run("refuses a hard (retryClass never) decline outright, regardless of attempt count", func(t *testing.T) {
		result := CanRetrySameInstrument(DeclineForRetryCheck(domain.RetryClassNever), 0, DefaultRetryPolicy)
		if result.Allowed {
			t.Fatal("expected refused")
		}
		if !regexp.MustCompile(`(?i)hard decline`).MatchString(result.Reason) {
			t.Fatalf("reason %q does not mention hard decline", result.Reason)
		}
	})

	t.Run(`refuses a "review" (unmapped) decline outright`, func(t *testing.T) {
		result := CanRetrySameInstrument(DeclineForRetryCheck(domain.RetryClassReview), 0, DefaultRetryPolicy)
		if result.Allowed {
			t.Fatal("expected refused")
		}
		if !regexp.MustCompile(`(?i)review`).MatchString(result.Reason) {
			t.Fatalf("reason %q does not mention review", result.Reason)
		}
	})

	t.Run("allows a retryable decline under the attempt cap", func(t *testing.T) {
		result := CanRetrySameInstrument(DeclineForRetryCheck(domain.RetryClassSameInstrumentLater), 1, DefaultRetryPolicy)
		if !result.Allowed {
			t.Fatalf("expected allowed, got refused: %s", result.Reason)
		}
	})

	t.Run("refuses a retryable decline once the attempt cap is reached", func(t *testing.T) {
		result := CanRetrySameInstrument(
			DeclineForRetryCheck(domain.RetryClassSameInstrumentLater),
			DefaultRetryPolicy.MaxAttemptsPerPayment,
			DefaultRetryPolicy,
		)
		if result.Allowed {
			t.Fatal("expected refused")
		}
		if !regexp.MustCompile(`(?i)max attempts`).MatchString(result.Reason) {
			t.Fatalf("reason %q does not mention max attempts", result.Reason)
		}
	})

	t.Run("a hard decline is refused even with zero prior attempts and plenty of budget left", func(t *testing.T) {
		// Guards against a bug where the attempt-cap check runs first and
		// masks the hard-decline check.
		result := CanRetrySameInstrument(DeclineForRetryCheck(domain.RetryClassNever), 0, RetryPolicyConfig{
			MaxAttemptsPerPayment: 100,
			MinSpacing:            0,
		})
		if result.Allowed {
			t.Fatal("expected refused")
		}
	})

	t.Run("the 4th attempt (attemptsSoFar == 3) is refused under the default policy", func(t *testing.T) {
		result := CanRetrySameInstrument(DeclineForRetryCheck(domain.RetryClassSameInstrumentLater), 3, DefaultRetryPolicy)
		if result.Allowed {
			t.Fatal("expected the 4th attempt to be refused (max 3 per payment)")
		}
	})

	t.Run("retry_after_challenge and retry_different_psp are retryable (subject to the attempt cap)", func(t *testing.T) {
		if !CanRetrySameInstrument(DeclineForRetryCheck(domain.RetryClassRetryAfterChallenge), 0, DefaultRetryPolicy).Allowed {
			t.Fatal("expected retry_after_challenge to be retryable")
		}
		if !CanRetrySameInstrument(DeclineForRetryCheck(domain.RetryClassRetryDifferentPSP), 0, DefaultRetryPolicy).Allowed {
			t.Fatal("expected retry_different_psp to be retryable")
		}
	})
}

func TestHasSufficientSpacing(t *testing.T) {
	t.Run("is false immediately after the last attempt", func(t *testing.T) {
		now := time.Date(2026, 1, 1, 0, 0, 0, 0, time.UTC)
		if HasSufficientSpacing(now, now, DefaultRetryPolicy) {
			t.Fatal("expected insufficient spacing")
		}
	})

	t.Run("is true once minSpacing has elapsed", func(t *testing.T) {
		last := time.Date(2026, 1, 1, 0, 0, 0, 0, time.UTC)
		now := last.Add(DefaultRetryPolicy.MinSpacing)
		if !HasSufficientSpacing(last, now, DefaultRetryPolicy) {
			t.Fatal("expected sufficient spacing")
		}
	})

	t.Run("is false one millisecond before minSpacing has elapsed", func(t *testing.T) {
		last := time.Date(2026, 1, 1, 0, 0, 0, 0, time.UTC)
		now := last.Add(DefaultRetryPolicy.MinSpacing - time.Millisecond)
		if HasSufficientSpacing(last, now, DefaultRetryPolicy) {
			t.Fatal("expected insufficient spacing")
		}
	})
}
